design/ceph/osd-migration.md

```diff
@@ -0,0 +1,227 @@
+- Once the device is cleaned up, its ready to be used again for OSD. Ensure that same OSD ID is used to prepare the OSD again. This cane be done by passing the OSD ID as `--osd-id` flag.
+- Prepare the OSD using the same `ROOK_REPLACE_OSD_ID` and `ROOK_REPLACE_OSD_PATH` using following commands:
+
+    ```
+    ceph-volume lvm prepare --osd-id {ROOK_REPLACE_OSD_ID} --data {ROOK_REPLACE_OSD_PATH}
+    ```
+
+#### Manage OSD migration status
+
+- OSD migration status will managed by `OSDReplaceConfig` API. This will be stored as the config map. For example:
+``` yaml oc get cm -n rook-ceph osd-replace-config -o yaml
+apiVersion: v1
+data:
+  config: '{"osdList":{"Q":[{"id":1,"path":"/mnt/set1-data-1qwdpk","node":"minikube-m03"},{"id":2,"path":"/mnt/set1-data-2c8h9k","node":"minikube-m04"}]},"isPgHealthy":true,"pgHealthMessage":"all
+    PGs in cluster are clean","lastReplacedOSD":{"id":0,"path":"/mnt/set1-data-045k6q","node":"minikube-m02"},"lastReplacedTimeStamp":"2023-07-18T14:42:29.541498777Z","startTimeStamp":"2023-07-18T14:41:33.180909077Z"}'
+kind: ConfigMap
+metadata:
+  creationTimestamp: "2023-07-18T14:42:29Z"
+  name: osd-replace-config
+  namespace: rook-ceph
+  ownerReferences:
+  - apiVersion: ceph.rook.io/v1
+    blockOwnerDeletion: true
+    controller: true
+    kind: CephCluster
+    name: rook-ceph
+    uid: e634abc3-39fa-4dd0-a015-753030e64f95
+  resourceVersion: "3580"
+  uid: 7b7d35d4-7807-4286-b185-3f8562cb3cbe
+```
+- This config map will have the `ownerReference` of cephCluster resource.
+
+##### Ceph Cluster Status:
+
+- Presence of this config map `osd-replace-config` will indicate that OSD migration is in progress.
+- Ceph cluster status should also be set to `Progressing` while `osd-replace-config` is available.
+- The status `message` will have some useful information about OSDs that are pending migration.
+  ``` yaml
+  conditions:
+    - lastHeartbeatTime: "2023-07-19T05:24:19Z"
+      lastTransitionTime: "2023-07-19T05:21:30Z"
+      message: Cluster created successfully
+      reason: ClusterCreated
+      status: "True"
+      type: Ready
+    - lastHeartbeatTime: "2023-07-19T05:24:19Z"
+      lastTransitionTime: "2023-07-19T05:24:19Z"
+      message: '2 osds awaiting migration to "bluestore-rdr". PG Status: "cluster
+        is not fully clean. PGs: [{StateName:undersized+peered Count:1}]"'
+      reason: ClusterProgressing
+      status: "True"
+      type: Progressing
+    message: '2 osds awaiting migration to "bluestore-rdr". PG Status: "cluster is
+      not fully clean. PGs: [{StateName:undersized+peered Count:1}]"'
+    observedGeneration: 3
+    phase: Progressing
+    state: Creating
+    storage:
+      deviceClasses:
+      - name: hdd
+    version:
+      image: quay.io/guits/ceph-volume:bs-rdr
+      version: 17.2.6-373
+      ```
+
```

Comment:
```suggestion

## Planning

This feature is big. Going in stages will be best. Can break down stages:
* new OSDs (greenfield)
* existing OSDs on PVC without metadata devices
* existing OSDs on PVC with metadata devices
* existing node-based OSDs (harder because may have been set up using batch)

```

Reply:
added a new section for `Planning`.